File: Excel/900-999/957/957 Challenge.py
import pandas as pd
import re

path = "900-999/957/957 Split.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=10, skiprows=0)
test = pd.read_excel(path, usecols="C", nrows=10, skiprows=0)

def split_string(data, rules):
    if pd.isna(rules) or rules == "(none)":
        return data
    tokens = [t.strip() for t in rules.split(",")]
    cut_points = set()
    for tok in tokens:
        if tok.startswith("@"):
            fn = {"@DIGIT": str.isdigit, "@UPPER": str.isupper, "@ALPHA": str.isalpha}.get(tok)
            if fn:
                cut_points.update(i for i, ch in enumerate(data) if fn(ch))
        else:
            pos = int(tok)
            cut_points.add(pos)
    cut_points.update([0, len(data)])
    cuts = sorted(cut_points)
    parts = [data[cuts[i]:cuts[i+1]] for i in range(len(cuts)-1)]
    parts = [p for p in parts if p]
    
    return " | ".join(parts)

input["Answer"] = input.apply(lambda row: split_string(row["Data"], row["Split_Rules"]), axis=1)
print(input["Answer"].equals(test["Answer Expected"]))
# True 

File: Excel/900-999/957/957 Challenge.R
library(tidyverse)
library(readxl)

path <- "900-999/957/957 Split.xlsx"
input <- read_excel(path, range = "A1:B10")
test <- read_excel(path, range = "C1:C10")

split_string <- function(data, rules) {
  if (is.na(rules) || rules == "(none)") {
    return(data)
  }
  tokens <- str_split(rules, ",")[[1]] %>% str_trim()
  cut_points <- c()
  for (tok in tokens) {
    if (str_starts(tok, "@")) {
      chars <- str_split(data, "")[[1]]
      pattern <- c(`@DIGIT` = "\\d", `@UPPER` = "[A-Z]", `@ALPHA` = "[A-Za-z]")[tok]
      if (!is.na(pattern)) {
        cut_points <- c(cut_points, which(str_detect(chars, pattern)) - 1)
      }
    } else {
      pos <- as.integer(tok)
      cut_points <- c(cut_points, pos)
    }
  }
  all_cuts <- sort(unique(c(0, cut_points, nchar(data))))
  parts <- map2_chr(
    all_cuts[-length(all_cuts)],
    all_cuts[-1],
    ~ substr(data, .x + 1, .y)
  )
  parts <- parts[parts != ""]
  paste(parts, collapse = " | ")
}
result <- input %>%
  mutate(Answer = map2_chr(Data, Split_Rules, split_string))

all.equal(result$Answer, test$`Answer Expected`)
# [1] TRUE
